azurelinuxagent/common/osutil/default.py

```diff
@@ -1313,28 +1313,75 @@ def device_for_ide_port(self, port_id):
 
     def set_hostname_record(self, hostname):
         fileutil.write_file(conf.get_published_hostname(), contents=hostname)
+        # If cloud init hostname file exists, update the file with the hostname. During service initialization, if
+        # provisioning was done by cloud-init the hostname will be set to whatever value is in the cloud init hostname
+        # file, so we should keep it updated with each hostname change.
+        if os.path.exists(self._get_cloud_init_hostname_path()):
+            self._set_cloud_init_hostname(hostname)
```

Comment:
Keep the most updated hostname in set-hostname (/var/lib/cloud/data/set-hostname), if the file exists. 

This change was made because the agent gets the hostname at every service initialization. So if set-hostname is not updated with the current hostname, the agent will revert to the original hostname at provisioning during service initialization.

Reply:
we shouldn't modify this file, since it was not created by the agent